synapse/handlers/oidc.py

```diff
@@ -935,17 +954,38 @@ async def handle_redirect_request(
 
         state = generate_token()
         nonce = generate_token()
+        code_verifier = ""
 
         if not client_redirect_url:
             client_redirect_url = b""
 
+        # If code challenge is supported and configured, use it.
+        extra_grant_values = {}
+        if self._code_challenge_method is not None:
+            code_verifier = generate_token(48)
+
+            # Default to the values for plain.
+            extra_grant_values = {"code_challenge_method": self._code_challenge_method}
+
+            if self._code_challenge_method == "S256":
+                extra_grant_values["code_challenge"] = create_s256_code_challenge(
+                    code_verifier
+                )
+            elif self._code_challenge_method == "plain":
+                extra_grant_values["code_challenge"] = code_verifier
```

Comment:
I wouldn't bother supporting the `plain` challenge. This only exists for constrained client, which is not our case.
This is what the [RFC says](https://www.rfc-editor.org/rfc/rfc7636.html#section-4.2):

```
   If the client is capable of using "S256", it MUST use "S256", as
   "S256" is Mandatory To Implement (MTI) on the server.  Clients are
   permitted to use "plain" only if they cannot support "S256" for some
   technical reason and know via out-of-band configuration that the
   server supports "plain".
```

Reply:
I removed support for this in 5e910b8d76e539fb9e2c02799cd9c258395f9c96.